src/passes/J2CLItableMerging.cpp

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This is a J2CL specific pass that merges itables into vtables. It is meant
+// to be run at the beginning before structs corresponding to Java classes are
+// optimized.
+//
+// Embeds itables into vtables to reduce memory usage.
+
+#include <unordered_map>
+#include <unordered_set>
+
+#include "ir/effects.h"
+#include "ir/localize.h"
+#include "ir/ordering.h"
+#include "ir/struct-utils.h"
+#include "ir/subtypes.h"
+#include "ir/type-updating.h"
+#include "ir/utils.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm-type.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+// Information about the structs that have vtables and itables.
+struct StructInfo {
+  HeapType javaClass;
+  HeapType itable;
+  HeapType vtable;
+
+  StructInfo(const HeapType& javaClass,
+             const HeapType& vtable,
+             const HeapType& itable)
+    : javaClass(javaClass), itable(itable), vtable(vtable) {}
+};
+
+struct J2CLItableMerging : public Pass {
+  // Keep track of all the structInfos so that they will be automatically
+  // released after the pass is done.
+  std::list<StructInfo> structInfos;
+
+  // Globals that hold vtables and itables indexed by their heap type.
+  // There is exactly 1 global for each vtable/itable type.
+  std::unordered_map<HeapType, std::unique_ptr<Global>*> tableGlobalsByType;
```

Comment:
```suggestion
  std::unordered_map<HeapType, Global*> tableGlobalsByType;
```
Just `Global*` should be enough. You can get the `Global*` from a unique_ptr by doing `.get()`.